### PR TITLE
drivers: gpio: expose gpio_utils.h to external GPIO drivers

### LINKED ISSUE #48609: drivers: gpio: expose gpio_utils.h to external GPIO drivers
**Is your enhancement proposal related to a problem? Please describe.**
The problem is with external gpio drivers. These helpers are needed but by convention everything that is not under "include" is "private" headers.

**Describe the solution you'd like**

- Remove inlines
- Put this code into drivers/gpio/gpio_util.c
- Put public interface under include/drivers/gpio/gpio_util.h
- Add a kconfig option which is enabled by default if GPIO is enabled

**Describe alternatives you've considered**
Moving the header into the repo with external drivers. Not optimal because it creates duplication. 

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by balajikulkarni**
Expose "gpio_utils.h" header for external GPIO drivers.

Fixes #48609.

Signed-off-by: Balaji Kulkarni <balaji.kulkarni92@gmail.com>

### CODE REVIEW

**Comment by balajikulkarni** [comment]
@nashif, @carlescufi, @galak could you please have a look at the PR or suggest any other reviewers who need to review it?

File: include/zephyr/drivers/gpio/gpio_utils.h [hunk 1]
```diff
@@ -4,9 +4,6 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-/**
- * @file Header where utility code can be found for GPIO drivers
- */
 
 #ifndef ZEPHYR_DRIVERS_GPIO_GPIO_UTILS_H_
 #define ZEPHYR_DRIVERS_GPIO_GPIO_UTILS_H_
```
Comment (cfriedt) [inline include/zephyr/drivers/gpio/gpio_utils.h:9]:
```suggestion
#define ZEPHYR_INCLUDE_DRIVERS_GPIO_GPIO_UTILS_H_
```

**Review by cfriedt — CHANGES_REQUESTED** [review]
@balajikulkarni - please also need to get rid of the merge commits. 
```
git rebase -i <commit to remove>^
```

**Comment by cfriedt** [comment]
@balajikulkarni - can you squash all of your changes into a single commit as well?
`git rebase -i HEAD~6` and then replace `pick` with `squash` for all but the first commit.

**Comment by balajikulkarni** [comment]
> @balajikulkarni - can you squash all of your changes into a single commit as well? `git rebase -i HEAD~6` and then replace `pick` with `squash` for all but the first commit.

Done.

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
@henrikbrixandersen - does this seem reasonable to you?

**Review by henrikbrixandersen — APPROVED** [review]
Yes.

**Pull request merged**